File: solution/day21/src/main/java/audit/Persister.java
package audit;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;

public class Persister {
    public FileContent[] readDirectory(String directory) throws IOException {
        return Files.walk(Paths.get(directory))
                .filter(Files::isRegularFile)
                .map(Persister::readFile)
                .toArray(FileContent[]::new);
    }

    public void applyUpdate(String directory, FileUpdate update) throws IOException {
        Files.writeString(Path.of(directory, update.fileName()), update.newContent());
    }

    private static FileContent readFile(Path f) {
        try {
            return new FileContent(f.getFileName().toString(), new ArrayList<>(Files.readAllLines(f)));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: solution/kotlin/day21/src/main/kotlin/audit/Persister.kt
package audit

import java.nio.file.Path
import kotlin.io.path.*

class Persister {
    fun readDirectory(directory: String): List<FileContent> =
        Path(directory)
            .listDirectoryEntries()
            .filter { it.isRegularFile() }
            .map { readFile(it) }

    fun applyUpdate(directory: String, update: FileUpdate) {
        Path(directory, update.fileName)
            .writeText(update.newContent)
    }

    private fun readFile(file: Path): FileContent =
        FileContent(file.name, file.readLines())
}

